crates/cargo-bsp/src/project_model/rust_extension/package.rs

```diff
@@ -0,0 +1,119 @@
+
+use crate::project_model::cargo_package::CargoPackage;
+use crate::project_model::metadata_edition_to_bsp_edition;
+use crate::project_model::rust_extension::{
+    find_node, get_nodes_from_metadata, target::metadata_targets_to_rust_extension_targets,
+};
+use crate::project_model::workspace::ProjectWorkspace;
+use crate::utils::uri::file_uri;
+use bsp_types::extensions::{RustFeature, RustPackage, RustPackageOrigin};
+use bsp_types::BuildTargetIdentifier;
+use std::collections::{HashMap, HashSet, VecDeque};
+
+fn resolve_origin(package: &mut RustPackage, workspace: &ProjectWorkspace) {
+    if workspace.is_package_part_of_workspace(&package.id) {
+        package.origin = RustPackageOrigin::Workspace;
+    } else {
+        package.origin = RustPackageOrigin::Dependency;
+    }
+}
+
+fn resolve_enabled_features(package: &mut RustPackage, nodes: &[cargo_metadata::Node]) {
+    if let Some(n) = find_node(
+        nodes,
+        &package.id,
+        "Proceeding with empty enabled features.",
+    ) {
+        package.enabled_features = n.features.clone()
+    }
+}
+
+fn metadata_features_to_rust_extension_features(
+    metadata_features: HashMap<String, Vec<String>>,
+) -> Vec<RustFeature> {
+    metadata_features
+        .into_iter()
+        .map(|(f, deps)| RustFeature {
+            name: f,
+            dependencies: deps,
+        })
+        .collect()
+}
+
+fn metadata_package_to_rust_extension_package(
+    metadata_package: cargo_metadata::Package,
+) -> RustPackage {
+    let all_targets = metadata_targets_to_rust_extension_targets(metadata_package.targets);
+    RustPackage {
+        id: metadata_package.id.clone().to_string(),
+        root_url: file_uri(metadata_package.manifest_path.parent().unwrap().to_string()),
+        name: metadata_package.name.clone(),
+        version: metadata_package.version.to_string(),
+        edition: metadata_edition_to_bsp_edition(metadata_package.edition),
+        source: metadata_package.source.map(|s| s.to_string()),
+        features: metadata_features_to_rust_extension_features(metadata_package.features),
+        // In our case targets = all_targets. This field is needed for Bazel
+        targets: all_targets.clone(),
+        all_targets,
+        // The rest of the fields is resolved later
+        ..RustPackage::default()
+    }
+}
+
+/// Returns a list of rust extension packages from which provided targets depend on
+pub fn get_rust_packages_related_to_targets(
+    workspace: &ProjectWorkspace,
+    metadata: &cargo_metadata::Metadata,
+    targets: &[BuildTargetIdentifier],
+) -> Vec<RustPackage> {
+    let target_related_packages_names: Vec<String> = targets
+        .iter()
+        .filter_map(|t| workspace.get_package_related_to_target(t))
+        .flat_map(|p| find_all_packages(p, &metadata.packages))
+        .collect::<HashSet<_>>()
+        .into_iter()
+        .collect();
+
+    let nodes = get_nodes_from_metadata(metadata);
```

Comment:
jak się usunie resolve_enabled_features(), to można też przenieść get_nodes do dependency,rs, gdyż aktualnie znajduje się w mod.rs